x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/ToCharFormatter.java

```diff
@@ -0,0 +1,376 @@
+        } else if (offset.matches("^[+-][0-9][0-9]:00$")) {
+            offset = offset.substring(0, offset.length() - 3);
+        }
+        return offset.substring(0, Math.min(offset.length(), 6));
+    }
+
+    private static String removeLeadingZerosFromOffset(String offset) {
+        if (offset.matches("[+-]0{1,2}")) {
+            return offset.substring(0, 2);
+        } else {
+            if (offset.startsWith("+0")) {
+                return "+" + offset.substring(2);
+            } else if (offset.startsWith("-0")) {
+                return "-" + offset.substring(2);
+            } else {
+                return offset;
+            }
+        }
+    }
+
+    private static String absoluteWeekBasedYear(TemporalAccessor t) {
+        int year = t.get(IsoFields.WEEK_BASED_YEAR);
+        year = year > 0 ? year : -(year - 1);
+        return String.format(Locale.ROOT, "%04d", year);
+    }
+
+    private static String firstDigitsOfNanos(String nano, int digits) {
+        return String.format(Locale.ROOT, "%09d", Integer.parseInt(nano)).substring(0, digits);
+    }
+
+    private static String lastNCharacter(String s, int n) {
+        return s.substring(Math.max(0, s.length() - n));
+    }
+
+    private static String zoneAbbreviationOf(TemporalAccessor temporalAccessor) {
+        String zone = ZoneId.from(temporalAccessor).getDisplayName(TextStyle.SHORT, Locale.ROOT);
+        return "Z".equals(zone) ? "UTC" : zone;
+    }
+
+    private static class Builder {
+
+        private final String pattern;
+        private boolean lowercaseAccepted = true;
+        private Function<String, String> fillMode = ToCharFormatter::removeLeadingZeros;
+        private boolean hasOrdinalSuffix = true;
+        private Function<TemporalAccessor, String> formatFn;
+
+        Builder(String pattern) {
+            this.pattern = pattern;
+        }
+
+        public Builder formatFn(final String javaPattern) {
+            return formatFn(javaPattern, null);
+        }
+
+        public Builder formatFn(final String javaPattern, final Function<String, String> additionalMapper) {
+            this.formatFn = temporalAccessor -> {
+                String formatted = DateTimeFormatter.ofPattern(javaPattern != null ? javaPattern : "'" + pattern + "'", Locale.ROOT)
+                    .format(temporalAccessor);
+                return additionalMapper == null ? formatted : additionalMapper.apply(formatted);
+            };
+            return this;
+        }
+
+        public Builder formatFn(Function<TemporalAccessor, String> formatFn) {
+            this.formatFn = formatFn;
+            return this;
+        }
+
+        public Builder noFillMode() {
+            this.fillMode = null;
+            return this;
+        }
+
+        public Builder text() {
+            return text(ToCharFormatter::removePadding);
+        }
+
+        public Builder text(Function<String, String> fillMode) {
+            this.hasOrdinalSuffix = false;
```

Comment:
Since there is a relationship between fillMode and ordinalSuffix should that apply to noFillMode as well?

Reply:
No. For example `FF6` (6 digit nanosecond) should not remove the leading zeros, but can have ordinal suffix (`000001st` nanosecond).